io/zenoh-links/zenoh-link-shm/src/unix/unicast.rs

```diff
@@ -0,0 +1,573 @@
+        self.get_w_mut().write_all(buffer).await
+    }
+
+    async fn read(&self, buffer: &mut [u8]) -> ZResult<usize> {
+        self.get_r_mut().read(buffer).await
+    }
+
+    async fn read_exact(&self, buffer: &mut [u8]) -> ZResult<()> {
+        self.get_r_mut().read_exact(buffer).await
+    }
+
+    #[inline(always)]
+    fn get_src(&self) -> &Locator {
+        &self.local
+    }
+
+    #[inline(always)]
+    fn get_dst(&self) -> &Locator {
+        &self.remote
+    }
+
+    #[inline(always)]
+    fn get_mtu(&self) -> u16 {
+        LINUX_PIPE_MAX_MTU
+    }
+
+    #[inline(always)]
+    fn is_reliable(&self) -> bool {
+        true
+    }
+
+    #[inline(always)]
+    fn is_streamed(&self) -> bool {
+        true
+    }
+}
+
+impl fmt::Display for UnicastPipe {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{} => {}", self.local, self.remote)?;
+        Ok(())
+    }
+}
+
+impl fmt::Debug for UnicastPipe {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Shm")
+            .field("src", &self.local)
+            .field("dst", &self.remote)
+            .finish()
+    }
+}
+
+pub struct LinkManagerUnicastPipe {
+    manager: Arc<NewLinkChannelSender>,
+    listeners: std::sync::RwLock<HashMap<EndPoint, UnicastPipeListener>>,
+}
+
+impl LinkManagerUnicastPipe {
+    pub fn new(manager: NewLinkChannelSender) -> Self {
+        Self {
+            manager: Arc::new(manager),
+            listeners: std::sync::RwLock::new(HashMap::new()),
+        }
+    }
+}
+impl ConstructibleLinkManagerUnicast<()> for LinkManagerUnicastPipe {
+    fn new(new_link_sender: NewLinkChannelSender, _: ()) -> ZResult<Self> {
+        Ok(Self::new(new_link_sender))
+    }
+}
+
+#[async_trait]
+impl LinkManagerUnicastTrait for LinkManagerUnicastPipe {
+    async fn new_link(&self, endpoint: EndPoint) -> ZResult<LinkUnicast> {
+        let pipe = UnicastPipeClient::connect_to(endpoint).await?;
+        Ok(LinkUnicast(Arc::new(pipe)))
+    }
+
+    async fn new_listener(&self, endpoint: EndPoint) -> ZResult<Locator> {
+        let listener = UnicastPipeListener::listen(endpoint.clone(), self.manager.clone()).await?;
+        let locator = listener.uplink_locator.clone();
+        self.listeners.write().unwrap().insert(endpoint, listener);
+        Ok(locator)
+    }
+
+    async fn del_listener(&self, endpoint: &EndPoint) -> ZResult<()> {
+        let removed = self.listeners.write().unwrap().remove(endpoint);
+        match removed {
+            Some(val) => {
+                val.stop_listening().await;
+                Ok(())
+            }
+            None => bail!("No listener found for endpoint {}", endpoint),
+        }
+    }
+
+    fn get_listeners(&self) -> Vec<EndPoint> {
+        self.listeners.read().unwrap().keys().cloned().collect()
+    }
+
+    fn get_locators(&self) -> Vec<Locator> {
+        self.listeners
+            .read()
+            .unwrap()
```

Comment:
The `unwrap()` might go away if the `RwLock` is from `async` instead of `std`. In any case, for style consistency, you might want to use `zread!` or `zasyncread!` macros.